File: clients/dart/cldf/lib/api/cldf_archive.dart
import '../models/manifest.dart';
import '../models/location.dart';
import '../models/sector.dart';
import '../models/route.dart';
import '../models/climb.dart';
import '../models/session.dart';
import '../models/tag.dart';
import '../models/media_item.dart';
import '../models/checksums.dart';

/// Represents a complete CLDF archive
class CLDFArchive {

  /// Creates a new [CLDFArchive] instance
  CLDFArchive({
    required this.manifest,
    required this.locations,
    this.sectors,
    this.routes,
    this.climbs,
    this.sessions,
    this.tags,
    this.mediaItems,
    this.checksums,
  });
  /// Archive metadata
  final Manifest manifest;

  /// Locations
  final List<Location> locations;

  /// Sectors
  final List<Sector>? sectors;

  /// Routes
  final List<Route>? routes;

  /// Climbs
  final List<Climb>? climbs;

  /// Sessions
  final List<Session>? sessions;

  /// Tags
  final List<Tag>? tags;

  /// Media items
  final List<MediaItem>? mediaItems;

  /// Checksums for files
  final Checksums? checksums;

  /// Check if the archive has routes
  bool get hasRoutes => routes != null && routes!.isNotEmpty;

  /// Check if the archive has climbs
  bool get hasClimbs => climbs != null && climbs!.isNotEmpty;

  /// Check if the archive has sessions
  bool get hasSessions => sessions != null && sessions!.isNotEmpty;

  /// Check if the archive has sectors
  bool get hasSectors => sectors != null && sectors!.isNotEmpty;

  /// Check if the archive has tags
  bool get hasTags => tags != null && tags!.isNotEmpty;

  /// Check if the archive has media
  bool get hasMedia => mediaItems != null && mediaItems!.isNotEmpty;
}

File: clients/java/cldf-java/src/main/java/app/crushlog/cldf/api/CLDFArchive.java
package app.crushlog.cldf.api;

import java.util.List;
import java.util.Map;

import app.crushlog.cldf.models.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Represents a complete CLDF (CrushLog Data Format) archive containing climbing data. This is the
 * main data structure for working with CLDF archives in memory.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CLDFArchive {

  private Manifest manifest;
  private Checksums checksums;
  private List<Location> locations;
  private List<Climb> climbs;
  private List<Session> sessions;
  private List<Route> routes;
  private List<Sector> sectors;
  private List<Tag> tags;
  private List<MediaMetadataItem> mediaItems;
  private Map<String, byte[]> mediaFiles;

  /**
   * Checks if this archive contains route data.
   *
   * @return true if routes are present and not empty
   */
  public boolean hasRoutes() {
    return routes != null && !routes.isEmpty();
  }

  /**
   * Checks if this archive contains sector data.
   *
   * @return true if sectors are present and not empty
   */
  public boolean hasSectors() {
    return sectors != null && !sectors.isEmpty();
  }

  /**
   * Checks if this archive contains tag data.
   *
   * @return true if tags are present and not empty
   */
  public boolean hasTags() {
    return tags != null && !tags.isEmpty();
  }

  /**
   * Checks if this archive contains media metadata.
   *
   * @return true if media items are present and not empty
   */
  public boolean hasMedia() {
    return mediaItems != null && !mediaItems.isEmpty();
  }

  /**
   * Checks if this archive contains embedded media files.
   *
   * @return true if media files are present and not empty
   */
  public boolean hasEmbeddedMedia() {
    return mediaFiles != null && !mediaFiles.isEmpty();
  }
}
